packages/api/authentication/handlers.go

```diff
@@ -0,0 +1,172 @@
+package authentication
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/gorilla/sessions"
+	"github.com/jiulongw/siwe-go"
+)
+
+var Store = sessions.NewCookieStore([]byte(os.Getenv("SESSION_KEY")))
+
+// seconds
+const MAX_NONCE_AGE = 30
+
+func IsAuthenticated(session *sessions.Session) bool {
+	return session.Values["siwe"] != nil
+}
+
+// Generates a nonce for the session
+func NonceHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		session, err := Store.Get(r, "nonce")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		nonce := siwe.GenerateNonce()
+
+		session.Values["nonce"] = nonce
+		// user has [MAX_NONCE_AGE] seconds to use nonce until it's not valid anymore
+		session.Options.MaxAge = MAX_NONCE_AGE
+		if err := session.Save(r, w); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(200)
+		// send back nonce + its max age
+		// [nonce] [max-age]
+		w.Write([]byte(fmt.Sprintf("%s %d", nonce, session.Options.MaxAge)))
+	}
+}
+
+// Validates signed payload with nonce
+func LoginHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var body LoginBody
+		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		siweMessage, err := siwe.MessageFromString(body.Message)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// parse hex signature into a sequence of bytes
+		// ignore 0x if starting with it
+		var signature []byte
+		if body.Signature[0:2] != "0x" {
```

Comment:
hm do we need to handle this case? there is also a `hexutil` library in `go-ethereum` which is a bit more ergonomic